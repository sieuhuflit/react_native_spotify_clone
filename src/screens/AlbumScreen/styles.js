import { StyleSheet } from 'react-native';
import { colors, device } from '../../constants';

const styles = StyleSheet.create({
  containerHeader: {
    height: 89,
    position: 'absolute',
    top: 0,
    width: '100%',
    zIndex: 100,
  },
  headerLinear: {
    height: 89,
    width: '100%',
  },
  header: {
    alignItems: 'center',
    flexDirection: 'row',
    justifyContent: 'space-between',
    paddingHorizontal: 24,
    paddingTop: device.iPhoneX ? 48 : 24,
    position: 'absolute',
    top: 0,
    width: '100%',
  },
  headerTitle: {
    fontSize: 16,
    color: colors.white,
    paddingHorizontal: 8,
    marginTop: 2,
    textAlign: 'center',
    width: device.width - 100,
  },
  containerFixed: {
    alignItems: 'center',
    paddingTop: device.iPhoneX ? 94 : 60,
    position: 'absolute',
    width: '100%',
  },
  containerLinear: {
    position: 'absolute',
    top: 0,
    width: '100%',
    zIndex: device.web ? 5 : 0,
  },
  containerImage: {
    shadowColor: colors.black,
    shadowOffset: { height: 8, width: 0 },
    shadowOpacity: 0.8,
    shadowRadius: 6,
    zIndex: device.web ? 20 : 0,
  },
  image: {
    height: 148,
    marginBottom: device.web ? 0 : 16,
    width: 148,
  },
  containerTitle: {
    marginTop: device.web ? 8 : 0,
    zIndex: device.web ? 20 : 0,
  },
  title: {
    fontSize: 20,
    fontWeight: 'bold',
    color: colors.white,
    paddingHorizontal: 24,
    marginBottom: 8,
    textAlign: 'center',
  },
  containerAlbum: {
    zIndex: device.web ? 20 : 0,
  },
  albumInfo: {
    fontSize: 12,
    color: colors.greyInactive,
    marginBottom: 48,
  },
  containerScroll: {
    paddingTop: 89,
  },
  containerSticky: {
    marginTop: device.iPhoneX ? 238 : 194,
  },
  containerShuffle: {
    alignItems: 'center',
    height: 50,
    shadowColor: colors.blackBg,
    shadowOffset: { height: -10, width: 0 },
    shadowOpacity: 0.2,
    shadowRadius: 20,
  },
  containerStickyLinear: {
    top: 0,
    position: 'absolute',
    width: '100%',
  },
  btn: {
    backgroundColor: colors.brandPrimary,
    borderRadius: 25,
    height: 50,
    width: 220,
  },
  btnText: {
    fontSize: 16,
    color: colors.white,
    letterSpacing: 1,
    textTransform: 'uppercase',
  },
  containerSongs: {
    alignItems: 'center',
    backgroundColor: colors.blackBg,
    minHeight: 540,
  },
  row: {
    alignItems: 'center',
    flexDirection: 'row',
    justifyContent: 'space-between',
    padding: 16,
    width: '100%',
  },
  downloadText: {
    fontSize: 18,
    fontWeight: 'bold',
    color: colors.white,
  },
});

export default styles;
