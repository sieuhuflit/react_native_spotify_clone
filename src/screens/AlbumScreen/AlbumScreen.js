import React from 'react';
import PropTypes from 'prop-types';
import { Animated, Alert, Image, Text, View } from 'react-native';
import Feather from 'react-native-vector-icons/Feather';
import { get } from 'lodash';
import { colors, device, globalStyles } from '../../constants';
import LinearGradient from '../../components/LinearGradient';
import LineItemSong from '../../components/LineItemSong';
import TouchIcon from '../../components/TouchIcon';
import TouchText from '../../components/TouchText';

import styles from './styles';

class AlbumScreen extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      scrollY: new Animated.Value(0),
    };
  }

  componentDidMount() {
    const { navigation, fetchAlbum } = this.props;
    const albumId = navigation.getParam('id');
    fetchAlbum(albumId);
  }

  componentWillUnmount() {
    const { resetAlbumDetail } = this.props;
    resetAlbumDetail();
  }

  changeSong = songData => {
    const {
      screenProps: { changeSong },
      albumDetail,
    } = this.props;
    changeSong(songData, albumDetail);
  };

  render() {
    const { navigation, albumDetail, screenProps } = this.props;
    const { scrollY } = this.state;
    if (albumDetail === null) {
      return <View style={[globalStyles.container, globalStyles.flexCenter]} />;
    }
    const { currentSongData } = screenProps;
    const albumName = get(albumDetail, 'name', '');
    const releaseDate = get(albumDetail, 'release_date', '');
    const artistName = get(albumDetail, 'artist[0].name', '');
    const albumImage = get(albumDetail, 'images[0].url', '');

    const stickyArray = device.web ? [] : [0];
    const headingRange = device.web ? [140, 200] : [230, 280];
    const shuffleRange = device.web ? [40, 80] : [40, 80];

    const opacityHeading = scrollY.interpolate({
      inputRange: headingRange,
      outputRange: [0, 1],
      extrapolate: 'clamp',
    });

    const opacityShuffle = scrollY.interpolate({
      inputRange: shuffleRange,
      outputRange: [0, 1],
      extrapolate: 'clamp',
    });

    return (
      <View style={globalStyles.container}>
        <View style={styles.containerHeader}>
          <Animated.View style={[styles.headerLinear, { opacity: opacityHeading }]}>
            <LinearGradient fill="#363230" height={89} />
          </Animated.View>
          <View style={styles.header}>
            <TouchIcon
              icon={<Feather color={colors.white} name="chevron-left" />}
              onPress={() => navigation.goBack(null)}
            />
            <Animated.View style={{ opacity: opacityShuffle }}>
              <Text style={styles.headerTitle}>{albumName}</Text>
            </Animated.View>
            <TouchIcon
              icon={<Feather color={colors.white} name="more-horizontal" />}
              onPress={() => Alert.alert('Setting')}
            />
          </View>
        </View>

        <View style={styles.containerFixed}>
          <View style={styles.containerLinear}>
            <LinearGradient fill="#363230" />
          </View>
          <View style={styles.containerImage}>
            <Image source={{ uri: albumImage }} style={styles.image} />
          </View>
          <View style={styles.containerTitle}>
            <Text ellipsizeMode="tail" numberOfLines={1} style={styles.title}>
              {albumName}
            </Text>
          </View>
          <View style={styles.containerAlbum}>
            <Text style={styles.albumInfo}>{`Album by ${artistName} Â· ${releaseDate}`}</Text>
          </View>
        </View>

        <Animated.ScrollView
          onScroll={Animated.event([{ nativeEvent: { contentOffset: { y: scrollY } } }], {
            useNativeDriver: true,
          })}
          scrollEventThrottle={16}
          showsVerticalScrollIndicator={false}
          stickyHeaderIndices={stickyArray}
          style={styles.containerScroll}
        >
          <View style={styles.containerSticky}>
            <Animated.View style={[styles.containerStickyLinear, { opacity: opacityShuffle }]}>
              <LinearGradient fill={colors.black20} height={50} />
            </Animated.View>
            <View style={styles.containerShuffle}>
              <TouchText
                onPress={() => null}
                style={styles.btn}
                styleText={styles.btnText}
                text="Shuffle Play"
              />
            </View>
          </View>
          <View style={styles.containerSongs}>
            {albumDetail.tracks &&
              albumDetail.tracks.items &&
              albumDetail.tracks.items.map((track, index) => (
                <LineItemSong
                  active={currentSongData && currentSongData === track.name}
                  key={index.toString()}
                  onPress={this.changeSong}
                  data={track}
                />
              ))}
          </View>
          <View style={globalStyles.spacer16} />
        </Animated.ScrollView>
      </View>
    );
  }
}

AlbumScreen.propTypes = {
  navigation: PropTypes.object.isRequired,
  screenProps: PropTypes.object.isRequired,
  fetchAlbum: PropTypes.func.isRequired,
  resetAlbumDetail: PropTypes.func.isRequired,
  albumDetail: PropTypes.object,
};

AlbumScreen.defaultProps = {
  albumDetail: null,
};

export default AlbumScreen;
