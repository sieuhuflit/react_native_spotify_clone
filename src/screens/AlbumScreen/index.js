import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import AlbumScreen from './AlbumScreen';
import { fetchAlbum, resetAlbumDetail } from '../../data/home/actions';
import HomeSelectors from '../../data/home/selectors';

const mapStateToProps = state => ({
  albumDetail: HomeSelectors.getAlbumDetail(state)
});

const mapDispatchToProps = dispatch =>
  bindActionCreators({ fetchAlbum, resetAlbumDetail }, dispatch);

export default connect(mapStateToProps, mapDispatchToProps)(AlbumScreen);
