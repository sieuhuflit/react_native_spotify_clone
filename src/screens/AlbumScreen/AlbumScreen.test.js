import React from 'react';
import { shallow } from 'enzyme';
import AlbumScreen from './AlbumScreen';

const setup = propOverrides => {
  const props = Object.assign(
    {
      fetchAlbum: jest.fn(),
      currentSongData: {
        id: 1,
        name: 'song name'
      },
      resetAlbumDetail: jest.fn(),
      screenProps: {
        changeSong: jest.fn()
      },
      navigation: {
        navigate: jest.fn(),
        getParam: jest.fn()
      },
      albumDetail: {
        id: 1
      }
    },
    propOverrides
  );
  const wrapper = shallow(<AlbumScreen {...props} />);
  return {
    props,
    wrapper
  };
};

describe('AlbumScreen check snapshot', () => {
  test('should match snapshot', () => {
    const { wrapper } = setup();
    expect(wrapper).toMatchSnapshot();
  });
});

describe('AlbumScreen check life cycle', () => {
  test('should componentDidMount work', () => {
    const { wrapper, props } = setup();
    wrapper.instance().componentDidMount();
    expect(props.fetchAlbum).toBeCalled();
  });
  test('should reset album detail work', () => {
    const { wrapper, props } = setup();
    wrapper.instance().componentWillUnmount();
    expect(props.resetAlbumDetail).toBeCalled();
  });
  test('should change instance song work', () => {
    const { wrapper, props } = setup();
    wrapper.instance().changeSong({ id: 1 });
    expect(props.screenProps.changeSong).toBeCalled();
  });
});
