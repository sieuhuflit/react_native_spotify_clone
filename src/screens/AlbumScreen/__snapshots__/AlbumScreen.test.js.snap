// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AlbumScreen check snapshot should match snapshot 1`] = `ShallowWrapper {}`;
